backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/policyunits/CpuPinningPolicyUnit.java

```diff
@@ -46,31 +61,66 @@ public CpuPinningPolicyUnit(PolicyUnit policyUnit,
+                // collect all pinned host cpus and merge them into one set
+                final Set<Integer> pinnedCpus = CpuPinningHelper.getAllPinnedPCpus(cpuPinning);
+                for (final VDS host : hosts) {
+                    var cpuTopology = resourceManager.getVdsManager(host.getId()).getCpuTopology();
+                    final Collection<Integer> onlineHostCpus = SlaValidator.getOnlineCpus(host);
+                    final Collection<Integer> dedicatedCpus = cpuTopology.stream().filter(VdsCpuUnit::isDedicated).map(VdsCpuUnit::getCpu).collect(Collectors.toList());
+                    final Collection availableCpus = CollectionUtils.subtract(onlineHostCpus, dedicatedCpus);
+                    final Collection difference = CollectionUtils.subtract(pinnedCpus, availableCpus);
+                    if (difference.isEmpty()) {
+                        candidates.add(host);
+                    } else {
+                        messages.addMessage(host.getId(), EngineMessage.VAR__DETAIL__VM_PINNING_PCPU_DOES_NOT_EXIST.name());
+                        messages.addMessage(host.getId(), String.format("$missingCores %1$s", StringUtils.join(difference, ", ")));
+                        log.debug("Host {} does not satisfy the cpu pinning constraints because of missing or offline cpus {}.",
+                                host.getId(),
+                                StringUtils.join(difference, ", "));
+                    }
+                }
+                break;
+            case DEDICATED:
+                for (final VDS host : hosts) {
+                    if (host.getCpuTopology() == null || host.getCpuTopology().isEmpty()) {
+                        // means CPU topology not reported for this host, lets ignore it
+                        messages.addMessage(host.getId(), EngineMessage.VAR__DETAIL__NO_HOST_DATA.name());
+                        log.debug("Host {} does not have the cpu topology data.", host.getId());
+                        continue;
+                    }
+
+                    Map<Guid, List<VdsCpuUnit>> vmToPendingDedicatedCpuPinnings =
+                            PendingCpuPinning.collectForHost(getPendingResourceManager(), host.getId());
+                    boolean isDedicatedCpuPinningPossibleAtHost =
+                            vdsCpuUnitPinningHelper.isDedicatedCpuPinningPossibleAtHost(
+                                    vmToPendingDedicatedCpuPinnings,
+                                    vm,
+                                    host);
+                    if (isDedicatedCpuPinningPossibleAtHost) {
+                        candidates.add(host);
+                    } else {
+                        messages.addMessage(host.getId(), EngineMessage.VAR__DETAIL__VM_PINNING_DEDICATED_NOT_FIT.name());
+                        log.debug("Host {} does not satisfy the cpu pinning constraints because of missing cpus to the" +
+                                        "needed topology.", host.getId());
```

Comment:
how about `Host {} does not satisfy CPU pinning constraints, cannot match virtual topology with available CPUs`?